File: client/services/auth.service.ts
import { z } from "zod/v3";
import { LoginRequest } from "@/types/payload/request/login.request";
import { ApiClient } from "@/services/api.client";
import { SuccessResponse } from "@/types/payload/response/common/success.response";
import { AuthResponse } from "@/types/payload/response/auth.response";
import { RegisterRequest } from "@/types/payload/request/register.request";
import { User } from "@/types/entity/user";

type LoginRequest = z.infer<typeof LoginRequest>;
type RegisterRequest = z.infer<typeof RegisterRequest>;

export class AuthService {
  static async login(
    data: LoginRequest,
  ): Promise<SuccessResponse<AuthResponse>> {
    return ApiClient.post("auth/login", {
      json: data,
    }).json<SuccessResponse<AuthResponse>>();
  }

  static async register(data: RegisterRequest): Promise<SuccessResponse<User>> {
    return ApiClient.post("auth/register", {
      json: data,
    }).json<SuccessResponse<User>>();
  }

  static async logout(): Promise<SuccessResponse<void>> {
    return ApiClient.post("auth/logout").json<SuccessResponse<void>>();
  }

  // static async refreshToken(): Promise<SuccessResponse<AuthResponse>> {
  //   return ApiClient.post("auth/refresh-token").json<
  //     SuccessResponse<AuthResponse>
  //   >();
  // }
}

File: backend/src/main/java/com/lerneon/backend/services/AuthService.java
package com.lerneon.backend.services;

import com.lerneon.backend.models.entity.User;
import com.lerneon.backend.models.payload.request.*;
import com.lerneon.backend.models.payload.response.AuthResponse;
import jakarta.servlet.http.HttpServletResponse;

public interface AuthService {
    AuthResponse login(HttpServletResponse response, LoginRequest loginRequest);

    User register(RegisterRequest registerRequest);

    void logout(HttpServletResponse response);

    User verifyUserAccount(OneTimePasswordRequest oneTimePasswordRequest);

    User verifyPasswordReset(OneTimePasswordRequest oneTimePasswordRequest);

    // FIXME: Move this to user service | 04/08/2025
    User changePassword(UpdatePasswordRequest updatePasswordRequest);

    User findUserByEmail(String email);
}
